### PR TITLE
RequestTestBase upgrade

### PR DESCRIPTION
**Pull request by fredg1**

### CODE REVIEW

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:118]: I'm not sure I understand the use case for all of these overloads.

You want to check that the path matches multiple regex patterns? and/or multiple strings? Are you trying to match multiple query parameters that may show up in an uncertain order?

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:43]: You shouldn't normally need to directly verify() your requests. Instead, prefer to use strict stubs (which are planned to be the default in Mockito 4).

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (fredg1) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java:66]: doesn't "BeforeAll" mean it will only run once (talking about ForbidNetworkAccess.beforeAll() )? Otherwise, it would be "BeforeEach"

Comment (fredg1) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java:66]: Also, no, `openNetwork` needs to be `BeforeEach`, as the factory needs to be refreshed between the tests (otherwise the stubs will carry between tests)

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (fredg1) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java:118]: The plan is to expect that people will want to match various parameters in URLs.

We don't want to force an "order" to those parameters.
For example,
```java
{"inventory.php"},
{"which=2"}
```
can match "inventory.php?which=2&pwd=\<pwd\>"
as well as "inventory.php?pwd=\<pwd\>&which=2"

Comment (heeheehee-kolmafia) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java]: I would prefer matchers for query parameters in that case.

For instance, something like:

    argThat(allOf(hasPath("inventory.php"), hasParam("which", 2)))

Although, even that is incorrect if you've got POST parameters that aren't part of the query string but are instead in the POST body.

Comment (heeheehee-kolmafia) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java]: In particular, checking for the substring "inventory.php" will incorrectly match a page "subinventory.php", and checking for "which=2" will incorrectly match "which=20" and "itemwhich=2".

Comment (fredg1) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java]: Mh... That seems possible + an interesting challenge.

This might take the rest of the day, at least. Is it possible to mark these as "changes requested" so that I can notify when done?

**Review by heeheehee-kolmafia — CHANGES_REQUESTED** [review]
requesting changes as requested

**Review by jaadams5 — DISMISSED** [review]
I am inclined to defer to @heeheehee-kolmafia for approval but if the requested changes have been resolved then I can merge.

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:193]: I'd prefer to call this something like "ParsedUrl" personally.

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:208]: nit: param.isEmpty()

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:206]: Instead of fullPath.substring(...), use fullPath.getQuery(). Similarly above, check that getQuery() != null instead of using indexOf("?").

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:198]: This won't work.

getPath() only returns the part of the path prior to the query and/or fragment.

**Review by heeheehee-kolmafia — CHANGES_REQUESTED** [review]

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:110]: nit: rename the variable name to parsedUrl?

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:235]: Why not store the String param and an ArgumentMatcher\<String> matcher?

That way instead of iterating through all the entries of params, you can use something like `matcher.matches(params.get(param))`.

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:91]: nit: "Sets ..."

"This is a fragment—a noun phrase or verb phrase, not a complete sentence. It does not begin with `A {@code Foo} is a...`, or `This method returns...`, nor does it form a complete imperative sentence like `Save the record.`. However, the fragment is capitalized and punctuated as if it were a complete sentence."

https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment

Ditto elsewhere.

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (heeheehee-kolmafia) [inline test/net/sourceforge/kolmafia/request/RequestTestBase.java:33]: nit: this should be in the body of the javadoc describing how to use it, not the description of what it is.

**Review by heeheehee-kolmafia — DISMISSED** [review]
I don't have much more input on this front.

I encourage anyone who's actually actively playing the game and/or engaged in active development of new features to weigh in regarding how they would want to use such a library.

@gausie @jaadams5

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (fredg1) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java:235]: I think I was still thinking about users maybe wanting to also submit a Pattern for the key.
I changed my mind mid-way, as the keys are pretty much specific and predictable enough that the user will know what to expect.

**Review by gausie — APPROVED** [review]
This is cool and a well-needed base for better testing in the project. It all looks good but this will be properly tested by an implementation. Thank you for writing it.

Will wait for heeheehee to approve your changes before merging

File: test/net/sourceforge/kolmafia/request/RequestTestBase.java
Comment (jaadams5) [reply test/net/sourceforge/kolmafia/request/RequestTestBase.java]: I think I agree with BeforeEach so am marking resolved.

**Review by jaadams5 — APPROVED** [review]
I'm going to approve with one comment.  @fredg1 please do what you can to make sure that this plays nicely with #220.  My life would be easier not having a merge  conflict or tests that stop playing nicely together.  I'll push tomorrow unless some other dev beats me to it.

**Comment by fredg1** [comment]
@jaadams5 do you want me to pull #220 into it, or do you think it may take longer?
There are indeed merge conflicts with it.

**Comment by jaadams5** [comment]
> @jaadams5 do you want me to pull #220 into it, or do you think it may take longer? There are indeed merge conflicts with it.

I am lazy.  If this gets committed first the I'm going to have to resolve merge conflicts in #220.  If this gets committed second I'm going to have to resolve merge conflicts because I press the button.

If there is an easy way for you to resolve the conflicts instead of me then I am in favor of it.  But if this one goes first there are at most two files to deal with so maybe I'm making a mountain out of a molehill?

**Comment by fredg1** [comment]
You could tell me when #220 is planned to be merged, and I could resolve the conflicts myself right after.

**Comment by jaadams5** [comment]
> You could tell me when #220 is planned to be merged, and I could resolve the conflicts myself right after.

That will work.  I will ping you when I close 220 and be prepared to reapprove and merge this.

**Comment by jaadams5** [comment]
@fredg1 OK - please resolve conflicts for me.  Thank you.

**Review by jaadams5 — APPROVED** [review]

**Pull request merged**
